meetings/working-groups/extensions/anonymous-extension-declarations.md

```diff
@@ -0,0 +1,368 @@
+
+``` c#
+public static class Bits
+{
+    extension(ref ulong) // underlying value is passed by ref
+    {
+        public bool this[int index]
+        {
+            get => (this & Mask(index)) != 0;
+            set => this = value ? this | Mask(index) : this & ~Mask(index); // mutates underlying value
+        }
+    }
+    static ulong Mask(int index) => 1ul << index;
+}
+```
+
+Underlying types in extension declarations can be or contain nullable reference types. If underlying types vary by nullability, separate extension declarations are needed.
+
+Individual extension members can place attributes on the implicit `this` parameter by using the `param` target in an attribute placed on the member itself. `param` is not currently allowed in this position, but is already used elsewhere in C# to refer to implicit parameters:
+
+``` c#
+public static class NullableExtensions
+{
+    extension(string?)
+    {
+        public string AsNotNull => this is null ? "" : this;
+        [param:NotNullWhen(false)] public bool IsNullOrEmpty => this is null or [];
+    }
+    extension(ref string?)
+    {
+        [param:NotNull] public void MakeNotNull() => this ??= "";
+    }
+    extension<T> (T) where T : class?
+    {
+        [param:NotNull] public void ThrowIfNull() => ArgumentNullException.ThrowIfNull(this);
+    }
+}
+```
+
+### Lowering
+
+Extension declarations give rise to static classes nested within the enclosing static class. These inner static classes are generated with unspeakable names and do not pollute the namespace of the enclosing static class. 
+
+Within a given enclosing static class, extension declarations with the same number of type parameters and equivalent constraints are merged into a single static class even if they differ on underlying type, ref-kind or type parameter names. Thus the minimum number of nested classes are generated to represent the combinations of type parameters and constraints present across all the extension declarations within the enclosing static class:
+
+``` c#
+public static class MyExtensions
+{
+    extension<TSource>(Span<TSource>) where TSource : class?
+    {
+        ...
+    }
+    extension<TElement>(Span<TElement?>) where TElement : class
+    {
+        ...
+    }
+}
+```
+
+Because the signatures are sufficiently equivalent, this is lowered to just one class:
+
+``` c#
+public static class MyExtensions
+{
+    public static class __E1<__T1> where __T1 : class
+    {
+        ...
+    }
+}
+```
+
+The requirement is not unlike that between the two parts of a partial method in today's C#.
+
+The members themselves are generated as static members. For instance members, insofar as it is possible to represent in IL, the same kind of member is generated, but with an extra first parameter for the underlying value. Where not possible, the body is represented as one or two (for accessors) static methods, and attributes are emitted to record their original member kind.
```

Comment:
It's certainly not possible for events at least, I believe. But if we're already in the territory of having to emit methods and to pair them up with attributes, why not go all the way and _only_ emit methods + supporting attributes?

Then, type parameters no longer ever need to be declared on the unspeakable nesting type, and then there would only need to be at most one unspeakable nesting type to contain all extensions.

This also opens the door to some or all extension members being declared as speakable static methods without being nested in an unspeakable type. If this is done merely for extension instance methods, it could be made to emit identically to a classic extension method. If this is done for all extension members, they could become callable from other languages and older compilers.